src/Bicep.Core/Parser/Parser.cs

```diff
@@ -57,45 +57,81 @@ public ProgramSyntax Program()
+                            LanguageConstants.OutputKeyword => this.OutputDeclaration(),
+                            _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
+                        },
+                        TokenType.NewLine => this.NewLine(),
 
-                return current.Type switch {
-                    TokenType.Identifier => current.Text switch {
-                        LanguageConstants.ParameterKeyword => this.ParameterDeclaration(),
-                        LanguageConstants.VariableKeyword => this.VariableDeclaration(),
-                        LanguageConstants.ResourceKeyword => this.ResourceDeclaration(),
-                        LanguageConstants.OutputKeyword => this.OutputDeclaration(),
                         _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                    },
-                    TokenType.NewLine => this.NewLine(),
-                    
-                    _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                };
-            }, consumeTerminator: false, TokenType.NewLine);
+                    };
+                },
+                RecoveryFlags.None,
+                TokenType.NewLine);
+
+        private RecoveryFlags CreateFlags(SyntaxBase precedingNode)
+        {
+            // local function
+            RecoveryFlags ConvertFlags(bool suppress) => suppress ? RecoveryFlags.SuppressDiagnostics : RecoveryFlags.None;
+
+            /*
+             * When we have an incomplete declarations like "param\n",
+             * the keyword is parsed but all other properties are set to 0-length SkippedTriviaSyntax or MalformedIdentifierSyntax
+             * to prevent stacking multiple parse errors on a 0-length span (which is technically correct but also confusing)
+             * we will only leave the first parse error
+             */
+            switch (precedingNode)
+            {
+                case MalformedIdentifierSyntax badIdentifier:
+                    return ConvertFlags(badIdentifier.Span.Length == 0);
+
+                case SkippedTriviaSyntax skipped:
+                    return ConvertFlags(skipped.Span.Length == 0);
+
+                default:
+                    return RecoveryFlags.None;
+            }
+        }
 
         private SyntaxBase ParameterDeclaration()
         {
             var keyword = ExpectKeyword(LanguageConstants.ParameterKeyword);
-            var name = Identifier(b => b.ExpectedParameterIdentifier());
-            var type = Type(b => b.ExpectedParameterType());
+            var name = this.IdentifierWithRecovery(b => b.ExpectedParameterIdentifier(), TokenType.Identifier, TokenType.NewLine);
+            var type = this.WithRecovery(() => Type(b => b.ExpectedParameterType()), CreateFlags(name), TokenType.Assignment, TokenType.LeftBrace, TokenType.NewLine);
```

Comment:
I'd be in favor of having CreateFlags return a bool and using a ternary to set the flags here. Otherwise I have to consider the possibility that CreateFlags may set the

Reply:
That is actually what I was trying to avoid. The ternary condition will be repeated everywhere with a different syntax node. We shouldn't be repeating ourselves that much.

Reply:
Renamed to `GetSuppressionFlag` to make it obvious that you may only get one flag here. To combine with other flags, the caller should just bitwise-OR the returned value with some other flag.